docgram/README.md

```diff
@@ -0,0 +1,185 @@
+
+* Local edits - edit rules that apply to the productions of a single non-terminal.
+If the non-terminal name isn't reserved, the rule is a local edit.  Local edits
+that begin with a different set of reserved names edit existing productions.  For
+example `binders: [ | DELETE Pcoq.Constr.binders ]` deletes the production
+`binders: [ | Pcoq.Constr.binders]`
+
+Productions that don't begin with a reserved name are added to the grammar,
+such as `empty: [ | ]`, which adds a new non-terminal `empty` with an
+empty production on the right-hand side.
+
+Another example: `LEFTQMARK: [ | "?" ]` is a local edit that treats `LEFTQMARK` as
+the name of a non-terminal and adds one production for it.  (We know that LEFTQMARK
+is a token but doc_gram does not.)  `SPLICE: [ | LEFTQMARK ]` requests replacing all
+uses of `LEFTQMARK` anywhere in the grammar with its productions and removing the
+non-terminal.  The combined effect of these two is to replace all uses of
+`LEFTQMARK` to `"?"`.
+
+Here are the current operations.  They are likely to be refined as we learn
+what operations are most useful while we update the mlg files and documentation:
+
+Global edits
+------------
+
+DELETE - deletes the specified non-terminals anywhere in the grammar.
+Useful for removing non-terminals that only do lookahead that shouldn't be
+in the documentation.
+
+RENAME - each production specifies an (old name, new name) pair of
+non-terminals to rename.
+
+SPLICE - requests replacing all uses of the nonterminals anywhere in the
+grammar with its productions and removing the non-terminal.  (Doesn't work
+recursively; splicing for both `A: [ | B ]` and `B: [ | C ]` has to be
+done in 2 SPLICE operations.)
+
+EXPAND_LISTS - expands LIST0, LIST1, LIST* ... SEP and OPT constructs into
+new non-terminals
+
+Local edits
+-----------
+
+DELETE <production> - removes the specified production from the grammar
+
+EDIT <production> - modifies the specified production using the following tags
+that are inserted in the production:
+  USE_NT <name> LIST* - extracts LIST* as new nonterminal with the specified
+    new non-terminal name
+  ADD_OPT <grammar symbol> - looks for a production that matches the specified
+    production **without** <grammar_symbol>.  If found, both productions are
+    replaced with single production with `OPT <grammar_symbol>`
+
+The current version handles a single USE_NT or ADD_OPT per EDIT.
+
+REPLACE <oldprod>
+WITH <newprod> (2 sequential productions) - removes <oldprod> and
+inserts <newprod> in its place.
+
+(any other NT name) - adds a new production (and possibly a new nonterminal)
+to the grammar.
+
+.rst file updates
+-----------------
+
+doc_gram updates the .rst file when it sees the following 3 lines
+
+```
+.. insertgram <start> <end>
+
+.. productionlist:: XXX
+```
+
+The end of the existing `productionlist` is recognized by a blank line or a line
+with `:name:` or `:undocumented:`
```

Comment:
Isn't the blank line case sufficient?

BTW, I'd like to see an example of a smaller `.. insertgram` comment in the manual (that only inserts a sub-part of the full grammar).

Reply:
No, doc_grammar needs to preserve `:name:` and `:undocumented:` since it doesn't have that information from the .mlg files.

> I'd like to see an example of a smaller `.. insertgram` comment in the manual

See the beginning of the added section.